// ---------------------------------------------------------------------------------------
//                                   Hiperspace
//                        Copyright (c) 2024 Cepheis Ltd
//                                    www.cepheis.com
//
// This file is part of Hiperspace and is distributed under the GPL Open Source License. 
// ---------------------------------------------------------------------------------------
syntax = "proto3";

option csharp_namespace = "Hiperspace.Remote";
import "google/protobuf/timestamp.proto";

package Hiperspace;

enum ResponseState
{
	Ok = 0;
	Skip = 1;
	Fail = 2;
	Idle = 3;	// the session was idle, and resource have been released
	Error = 4;	// another error (not related to comunication) has occured
}

enum HiperDrive
{
	Rocks = 0;
	Remote = 1;
}

message Token
{
	bytes TokenId = 1;
}

message OpenRequest
{
	Token Token = 1;
	string Path = 2;
	bytes MetaModel = 3;
	HiperDrive Driver = 4;
	bool Compression = 5;
	string Realm = 6;
	string UserName = 7;
}

message BindRequest
{
	bytes Key = 1;
	bytes Value = 2;
	Token Token = 3;
}
message BindVersionRequest
{
	bytes Key = 1;
	bytes Value = 2;
	google.protobuf.Timestamp Version = 3;
	Token Token = 4;
}

message BatchBindRequest
{
	repeated BindRequest Value = 1;
	Token Token = 2;
}
message BatchBindVersionRequest
{
	repeated BindVersionRequest Value = 1;
	Token Token = 3;
}

message KeyRequest
{
	bytes Key = 1;
	Token Token = 2;
}
message KeyVersionRequest
{
	bytes Key = 1;
	google.protobuf.Timestamp Version = 2;
	Token Token = 3;
}

message FindRequest 
{
	bytes Begin = 1;
	bytes End = 2;
	Token Token = 3;
}
message FindVersionRequest 
{
	bytes Begin = 1;
	bytes End = 2;
	google.protobuf.Timestamp Version = 3;
	Token Token = 4;
}

message Value
{
	bytes Content = 1;
	ResponseState State = 2;
}
message ValueVersion
{
	bytes Content = 1;
	google.protobuf.Timestamp Version = 2;
	ResponseState State = 3;
}

message KeyValue
{
	bytes Key = 1;
	bytes Value = 2;
	ResponseState State = 3;
}

message KeyValueVersion
{
	bytes Key = 1;
	bytes Value = 2;
	google.protobuf.Timestamp Version = 3;
	ResponseState State = 4;
}

message Values
{
	repeated KeyValue Content = 1;
	ResponseState State = 2;
}

message ValueVersions
{
	repeated KeyValueVersion Content = 1;
	ResponseState State = 2;
}

message VersionHistory
{
	repeated ValueVersion Content = 1;
	ResponseState State = 2;
}
// Provides a remote interface to a hiperspace driver
// for authorised users
service HiServ
{
	rpc Open (OpenRequest) returns (Token);
	rpc Close (Token) returns (Token);

	rpc Bind (BindRequest) returns (Value);
	rpc BatchBind (BatchBindRequest) returns (Values);
	rpc Find (FindRequest) returns (Values);
	rpc Get (KeyRequest) returns (Value);

	rpc BindVersion (BindVersionRequest) returns (ValueVersion);
	rpc BatchVersionBind (BatchBindVersionRequest) returns (Values);
	rpc FindVersion (FindVersionRequest) returns (ValueVersions);
	rpc GetVersion (KeyVersionRequest) returns (ValueVersion);
	rpc GetVersions (KeyRequest) returns (VersionHistory);
}